### ISSUE TITLE
SM4 in vpsm4-armv8.pl test failures on big-endian ARM processors

### ISSUE THREAD

**Issue by xu-yi-zhou** [openssl/openssl#19911]
SM4 in `vpsm4-armv8.pl` test failures on big-endian ARM processors, so I submitted a [PR](https://github.com/openssl/openssl/pull/19910) intending to fix it, but I'm not quite sure if it's a bug, correct me if I'm wrong. 

I used cross-compilation toolchain `aarch64_be-linux-gnu-gcc`, and the config like this:
```
./Configure linux-aarch64 --cross-compile-prefix=xxx/aarch64_be-linux-gnu- LDFLAGS=-mbig-endian
```

The following error is reported:
```
# ./evp_test evpciph_sm4.txt
1..1
    # Subtest: run_file_tests
    1..1
    # INFO:  @ test/testutil/stanza.c:21
    # Reading evpciph_sm4.txt
    # INFO:  @ test/testutil/stanza.c:122
    # Starting "SM4 test vectors from IETF draft-ribose-cfrg-sm4" tests at line 9
    # INFO:  @ test/evp_test.c:605
    # SM4-ECB is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-681edf34d206965e 86b3e94f536e4246
    # 0000:+f17b82d433defd17 0507ff454645ae72
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:11: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # INFO:  @ test/evp_test.c:605
    # SM4-ECB is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-681edf34d206965e 86b3e94f536e4246 681edf34d206965e 86b3e94f536e4246
    # 0000:+7319138cc31f4ebb ef0c58f0bd06dbd9 7319138cc31f4ebb ef0c58f0bd06dbd9
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-681edf34d206965e 86b3e94f536e4246 681edf34d206965e 86b3e94f536e4246
    # 0020:+7319138cc31f4ebb ef0c58f0bd06dbd9 7319138cc31f4ebb ef0c58f0bd06dbd9
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:16: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # INFO:  @ test/evp_test.c:605
    # SM4-CBC is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-2677f46b09c122cc 975533105bd4a22a f6125f7275ce552c 3a2bbcf533de8a3b
    # 0000:+28a932ad04946f68 c2eb7b01153671f5 03e9f76544196dda 4010a747197bc3c4
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:21: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/evp_test.c:605
    # SM4-CBC is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-2677f46b09c122cc 975533105bd4a22a f6125f7275ce552c 3a2bbcf533de8a3b
    # 0000:+ad32a928686f9404 017bebc2f5713615 65f7e903da6d1944 47a71040c4c37b19
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-fff5a4f208092c09 01ba02d577297736 9915e3fa2356c9f4 eb6460ecc457e7f8
    # 0020:+cccfecad16a2e080 800753e8fa9429e3 b0b7838162a8c79c f5fc18511a876c66
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0040:-e3cfa3deebfe9883 e3a48bcf7c4a11aa 3ec9e0d317c5d319 be72a5cdddec640c
    # 0040:+9ed776d32dbd4c73 c97a9a07f2941e57 5c9fa3b7a1f7323e cc437a01d0d16be5
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:27: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/evp_test.c:605
    # SM4-CBC is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-2677f46b09c122cc 975533105bd4a22a f6125f7275ce552c 3a2bbcf533de8a3b
    # 0000:+ad32a928686f9404 017bebc2f5713615 65f7e903da6d1944 47a71040c4c37b19
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-fff5a4f208092c09 01ba02d577297736 9915e3fa2356c9f4 eb6460ecc457e7f8
    # 0020:+cccfecad16a2e080 800753e8fa9429e3 b0b7838162a8c79c f5fc18511a876c66
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0040:-e3cfa3deebfe9883 e3a48bcf7c4a11aa 3ec9e0d317c5d319 be72a5cdddec640c
    # 0040:+d376d79e734cbd2d 079a7ac9571e94f2 b7a39f5c3e32f7a1 017a43cce56bd1d0
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0060:-6fc70bfa3ddaafff dd7c09b2774dcb2c ec29f0c6f0b6773e 985b3e395e924238
    # 0060:+c1795512080c80c1 588f863a5e368355 a24fea600b3ab13a c706ea3e597a1148
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0080:-505a8f120d9ca84d e5c3cf7e45f097b1 4b3a46c5b1068669 982a5c1f5f61be29
    # 0080:+b6579f19cdb5879e 4d90171458a8657d ee4954c9eafe1d32 825d2e1c417a8f37
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 00a0:-1b984f331d44ffb2 758f771672448fc9 57fa1416c446427a 41e25d5524a2418b
    # 00a0:+8273b89433499672 18ecabbff932651b 06b7b24f91769970 9ff9aae2bae6ffa0
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 00c0:-9d96b2f17582f0f1 aa9c204c6807f54f 7b6833c5f0085665 9ddabc245936868c
    # 00c0:+a323a2b2a0a0a93f 935f7b64df755e15 994a9a81156fc48e a8d7f35a9b5b1243
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:33: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/evp_test.c:605
    # SM4-OFB is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-693d9a535bad5bb1 786f53d7253a7056 f2075d28b5235f58 d50027e4177d2bce
    # 0000:+f058c7b3ba7530f8 fbdb45dd30119c62 f9fa55d09815854d a8a65399c6608458
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:39: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/evp_test.c:605
    # SM4-CFB is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-693d9a535bad5bb1 786f53d7253a7056 9ed258a85a0467cc 92aab393dd978995
    # 0000:+f058c7b3ba7530f8 fbdb45dd30119c62 3310de62b4e42c8f 229fc1ac3f0e9ebd
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:45: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/evp_test.c:605
    # SM4-CTR is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-c2b4759e78ac3cf4 3d0852f4e8d5f9fd 7256e8a5fcb65a35 0ee00630912e4449
    # 0000:+d9b3b92669b5e411 54b7e34b06bd6062 5dd539ba919570d5 7aa4cd5cac5342ad
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-2a0b17e1b85b060d 0fba612d8a958316 38b361fd5ffacd94 2f081485a83ca35d
    # 0020:+f9a6be384635d175 b5d4d61063bccfa3 05996086d9184c0c a16b12cc59a1b3a3
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:51: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/testutil/stanza.c:122
    # Starting "SM4 GCM test vectors from RFC8998" tests at line 57
    # INFO:  @ test/evp_test.c:605
    # SM4-GCM is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-17f399f08c67d5ee 19d0dc9969c4bb7d 5fd46fd375648906 9157b282bb200735
    # 0000:+1f4dd6a9f963c52d eb185ae4e1c079f4 83cccb8a5382e006 063d493455183527
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^   ^^^^^^^^^^^^^^^^
    # 0020:-d82710ca5c22f0cc fa7cbf93d496ac15 a56834cbcf98c397 b4024a2691233b8d
    # 0020:+824a6c453a3f4763 d7dec6ef84a74735 f62da8834cebccd0 5b66140ee8f3e6aa
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:59: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # INFO:  @ test/testutil/stanza.c:122
    # Starting "SM4 CCM test vectors from RFC8998" tests at line 67
    # INFO:  @ test/evp_test.c:605
    # SM4-CCM is fetched
    # INFO:  @ test/testutil/stanza.c:122
    # Starting "SM4 XTS test vectors from GB/T 17964-2021" tests at line 77
    # INFO:  @ test/evp_test.c:605
    # SM4-XTS is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-e9538251c71d7b80 bbe4483fef497bd1 2c5c581bd6242fc5 1e08964fb4f60fdb
    # 0000:+89c623c71de60f69 75555e29346f6d68 04fefce90de5fc20 466338e070bfb991
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-0ba42f6349927921 3d318d2c11f6886e 903be7f93a1b3479
    # 0020:+00a27a97924fc3bb 3d5273dc7ac873fc 2fbd6ece6009c8c2
    #       ^^^^^^^^^^^^^^^^   ^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:79: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/evp_test.c:605
    # SM4-XTS is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-e9538251c71d7b80 bbe4483fef497bd1 2c5c581bd6242fc5 1e08964fb4f60fdb
    # 0000:+89c623c71de60f69 75555e29346f6d68 04fefce90de5fc20 466338e070bfb991
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-0ba42f6349927921 3d318d2c11f6886e 903be7f93a1b3479
    # 0020:+00a27a97924fc3bb 3d5273dc7ac873fc 2fbd6ece6009c8c2
    #       ^^^^^^^^^^^^^^^^   ^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:85: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/testutil/stanza.c:122
    # Starting "SM4 XTS test vectors, while the XTS mode is standardized in IEEE Std 1619-2007" tests at line 92
    # INFO:  @ test/evp_test.c:605
    # SM4-XTS is fetched
    # ERROR: (memory) 'expected == got' failed @ test/evp_test.c:107
    # --- expected
    # +++ got
    # 0000:-e9538251c71d7b80 bbe4483fef497bd1 b3db1a3e60408c57 5d63ff7db39f8326
    # 0000:+89c623c71de60f69 75555e29346f6d68 00c9a653d0e100ad db4da9a9553f9685
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    # 0020:-0869f9e2585fec9f 0b863bf8fd784b86 27d16c0db6d2cfc7
    # 0020:+e7d57f9c0695a5db da5238ec98621018 6a5cd901fcd8b29e
    #       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
    #
    # INFO:  @ test/evp_test.c:3574
    # evpciph_sm4.txt:94: Source of above error (aligned in-place, not fragmented); unexpected error VALUE_MISMATCH
    # 0000007FA19F2010:error:07800081:common libcrypto routines:set_ptr_internal:param of incompatible type:crypto/params.c:1439:
    # INFO:  @ test/testutil/stanza.c:33
    # Completed 13 tests with 12 errors and 0 skipped
    # OPENSSL_TEST_RAND_ORDER=66413118
    not ok 1 - iteration 1
# OPENSSL_TEST_RAND_ORDER=66413118
not ok 1 - run_file_tests
```

**Comment by zorrorffm** [comment]
Thanks for the report and the fix, I will take a look.

**Comment by zorrorffm** [comment]
I guess this issue can be closed because #9910 has fixed it.
